File: MS_Order/src/main/java/tn/esprit/ms_order/services/OrderService.java
package tn.esprit.ms_order.services;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import tn.esprit.ms_order.entities.Order;
import tn.esprit.ms_order.entities.OrderItem;
import tn.esprit.ms_order.entities.OrderStatus;
import tn.esprit.ms_order.repositories.OrderRepository;

import java.util.List;

@Service
public class OrderService {

    private final OrderRepository orderRepository;
    public OrderService(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }

    public Order getOrCreateCart(Long userId) {
        return orderRepository
                .findByUserIdAndStatus(userId, OrderStatus.CART)
                .orElseGet(() -> {
                    Order cart = new Order();
                    cart.setUserId(userId);
                    cart.setStatus(OrderStatus.CART);
                    return orderRepository.save(cart);
                });
    }

    public Order addToCart(Long userId, Long articleId,
                           String articleName, Double price, int quantity) {
        Order cart = getOrCreateCart(userId);

        // Check if article already in cart, just update quantity
        cart.getItems().stream()
                .filter(i -> i.getArticleId().equals(articleId))
                .findFirst()
                .ifPresentOrElse(
                        i -> i.setQuantity(i.getQuantity() + quantity),
                        () -> {
                            OrderItem item = new OrderItem();
                            item.setOrder(cart);
                            item.setArticleId(articleId);
                            item.setArticleName(articleName);
                            item.setPrice(price);
                            item.setQuantity(quantity);
                            cart.getItems().add(item);
                        }
                );

        return orderRepository.save(cart);
    }

    public Order removeFromCart(Long userId, Long itemId) {
        Order cart = getOrCreateCart(userId);
        cart.getItems().removeIf(i -> i.getId().equals(itemId));
        return orderRepository.save(cart);
    }

    public Order confirmOrder(Long userId) {
        Order cart = orderRepository
                .findByUserIdAndStatus(userId, OrderStatus.CART)
                .orElseThrow(() -> new RuntimeException("No active cart for user: " + userId));

        if (cart.getItems().isEmpty())
            throw new RuntimeException("Cart is empty");

        // Later: call articleClient.reduceStock() for each item
        cart.setStatus(OrderStatus.CONFIRMED);
        return orderRepository.save(cart);
    }

    public List<Order> getOrdersByUser(Long userId) {
        return orderRepository.findByUserId(userId);
    }

    public Order getOrderById(Long orderId) {
        return orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found: " + orderId));
    }

    public Order updateStatus(Long orderId, OrderStatus status) {
        Order order = getOrderById(orderId);
        order.setStatus(status);
        return orderRepository.save(order);
    }

    public void cancelOrder(Long orderId) {
        Order order = getOrderById(orderId);
        if (order.getStatus() != OrderStatus.CART)
            throw new RuntimeException("Can only cancel orders in CART status");
        order.setStatus(OrderStatus.CANCELLED);
        orderRepository.save(order);
    }
}

File: src/app/services/order.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface OrderItem {
  id?: number;
  articleId: number;
  articleName: string;
  price: number;
  quantity: number;
}

export interface Order {
  id: number;
  userId: number;
  status: string;
  items: OrderItem[];
  total?: number;
  createdAt: string;
  updatedAt?: string;
}

@Injectable({
  providedIn: 'root'
})
export class OrderService {

  private baseUrl = 'http://localhost:8080/api/orders';

  constructor(private http: HttpClient) {}

  // Get all orders (for admin)
  getAllOrders(): Observable<Order[]> {
    return this.http.get<Order[]>(`${this.baseUrl}`);
  }

  // Get orders by user ID
  getOrdersByUser(userId: number): Observable<Order[]> {
    return this.http.get<Order[]>(`${this.baseUrl}/user/${userId}`);
  }

  // Get a specific order
  getOrderById(orderId: number): Observable<Order> {
    return this.http.get<Order>(`${this.baseUrl}/${orderId}`);
  }

  // Get cart
  getCart(userId: number): Observable<Order> {
    return this.http.get<Order>(`${this.baseUrl}/cart/${userId}`);
  }

  // Add to cart
  addToCart(userId: number, articleId: number, articleName: string, 
            price: number, quantity: number): Observable<Order> {
    const params = new HttpParams()
      .set('articleId', articleId.toString())
      .set('articleName', articleName)
      .set('price', price.toString())
      .set('quantity', quantity.toString());

    return this.http.post<Order>(
      `${this.baseUrl}/cart/${userId}/add`,
      {},
      { params }
    );
  }

  // Remove from cart
  removeFromCart(userId: number, itemId: number): Observable<Order> {
    return this.http.delete<Order>(`${this.baseUrl}/cart/${userId}/item/${itemId}`);
  }

  // Confirm order
  confirmOrder(userId: number): Observable<Order> {
    return this.http.post<Order>(`${this.baseUrl}/cart/${userId}/confirm`, {});
  }

  // Update order status
  updateStatus(orderId: number, status: string): Observable<Order> {
    const params = new HttpParams().set('status', status);
    return this.http.put<Order>(`${this.baseUrl}/${orderId}/status`, {}, { params });
  }

  // Cancel/Delete order
  cancelOrder(orderId: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${orderId}`);
  }
}
